src/contract_client.rs

```diff
@@ -668,18 +672,420 @@ impl ModuleDeployHandle {
     }
 }
 
+/// The outcome of invoking (dry-running) a transaction to update a smart
+/// contract instance. The variants describe the two cases of successfully
+/// simulating the transaction and rejecting the transaction due to some
+/// reverts.
+pub enum InvokeContractOutcome {
+    /// The simulation of the transaction was successful.
+    Success(SimulatedTransaction),
+    /// The transaction was rejected due to some reverts.
+    Failure(RejectedTransaction),
+}
+
+/// The `SimulatedTransaction` type is an alias for the `ContractUpdateBuilder`
+/// type. This type is used when an invoke (dry-run) of a transaction succeeds.
+/// This type includes a convenient send method to send and execute the
+/// transaction on-chain in a subsequent action. As such, it is a builder to
+/// simplify sending smart contract updates.
+pub type SimulatedTransaction = ContractUpdateBuilder;
+
+/// This type is used when an invoke (dry-run) of a transaction gets rejected.
+#[derive(Debug, Clone)]
+pub struct RejectedTransaction {
+    /// The return value of the transaction.
+    pub return_value:   Option<ReturnValue>,
+    /// The reject reason of the transaction.
+    pub reason:         RejectReason,
+    /// An optional human-readable decoded reason for the reject reason.
+    /// This is only available if the reject reason is a smart contract logical
+    /// revert and a valid error schema is available for decoding, or the
+    /// reject reason originates from the `concordium-std` crate.
+    /// The JSON value format of such errors is for example:
+    /// Object {\"Unauthorized\": Array []} or
+    /// Object {\"Custom\": Array [Object {\"Unauthorized\": Array []}]} (e.g.
+    /// the cis2_library error)
+    /// ...
+    pub decoded_reason: Option<Value>,
```

Comment:
I don't know if we should wrap this a new-type that we can attach logic to at a later point if we want? Or does that seem like overkill to you @limemloh @DOBEN ?